src/timeoffset.rs

```diff
@@ -0,0 +1,362 @@
+    pub fn from_qzsst_utc_time_of_week(
+        t_ref_nanos: (u32, u64),
+        polynomials: (f64, f64, f64),
+    ) -> Self {
+        Self::from_reference_time_of_week(
+            t_ref_nanos,
+            TimeScale::QZSST,
+            TimeScale::UTC,
+            polynomials,
+        )
+        .unwrap()
+    }
+
+    /// Returns both [TimeScale]s this [TimeOffset] allows converting to.
+    pub fn supported_timescales(&self) -> (TimeScale, TimeScale) {
+        (self.lhs, self.rhs)
+    }
+
+    /// Update this [TimeOffset] with new reference epoch and polynomials.
+    /// NB: this should be expressed in the left-hand side [TimeScale] and we have no means
+    /// to verify that.
+    pub fn update_mut(&mut self, t_ref_nanos: (u32, u64), polynomials: (f64, f64, f64)) {
+        self.t_ref_nanos = t_ref_nanos;
+        self.polynomials = polynomials;
+    }
+
+    /// Define a new [TimeOffset] with new reference [TimeScale], while preserving other components.
+    /// NB: this should be expressed in the left-hand side [TimeScale] and we do not verify it!
+    pub fn with_reference_timescale(&self, ts: TimeScale) -> Self {
+        let mut s = self.clone();
+        s.rhs = ts;
+        s
+    }
+
+    /// Define a new [TimeOffset] with new left-hand side [TimeScale], while preserving other components.
+    /// This needs to be coupled to either [Self::with_reference_time_of_week_nanos] or
+    /// [Self::with_reference_epoch] to remain correct.
+    pub fn with_lhs_timescale(&self, ts: TimeScale) -> Self {
+        let mut s = self.clone();
+        s.lhs = ts;
+        s
+    }
+
+    /// Define a new [TimeOffset] with new reference time of week (in nanoseconds), while preserving other components.
+    pub fn with_reference_time_of_week_nanos(&self, t_ref_nanos: (u32, u64)) -> Self {
+        let mut s = self.clone();
+        s.t_ref_nanos = t_ref_nanos;
+        s
+    }
+
+    /// Define a new [TimeOffset] with new reference [Epoch] with 1 ns precision.
+    pub fn with_reference_epoch(&self, t_ref: Epoch) -> Result<Self, TimeOffsetError> {
+        let mut s = self.clone();
+
+        if t_ref.time_scale != self.lhs {
+            return Err(TimeOffsetError::InvalidTimescale);
+        }
+
+        s.t_ref_nanos = t_ref.to_time_of_week();
+        Ok(s)
+    }
+
+    /// Define a new [TimeOffset] with new polynomials, while preserving other components.
+    pub fn with_polynomials(&self, polynomials: (f64, f64, f64)) -> Self {
+        let mut s = self.clone();
+        s.polynomials = polynomials;
+        s
+    }
+
+    /// Returns the total number of nanoseconds to apply to convert this [Epoch]
+    /// into either of [Self::supported_timescales].
+    /// NB:
+    /// - `t` must be expressed in either of [Self::supported_timescales].
+    /// - `t` should fall within the reference week, otherwise this will give invalid results.
+    pub fn time_correction_nanos(&self, t: Epoch) -> Result<f64, TimeOffsetError> {
```

Comment:
If `polynomials` use `Duration`, then most of this math becomes much simpler since `Duration` already handles the math.

Reply:
So Duration would only apply to the first term. And IMO it is important to not use Durations here, because we allow the user to not be limited to 1ns precision.

Reply:
In Nyx, I define a [`CommonPolynomial`](https://docs.rs/nyx-space/latest/nyx_space/polyfit/enum.CommonPolynomial.html) for maneuver definitions. I see a similarity here, which is why I proposed a polynomial structure.